<a href="https://colab.research.google.com/github/Rahma2015/Netflix-Movies-Simple-recommendation-System/blob/main/Netflix_Recommendation_System.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##EXPLORE DATA:

In [ ]:
!pip install opendatasets

In [ ]:
import opendatasets as od
import pandas as pd

In [ ]:

od.download(
    "https://www.kaggle.com/datasets/rishitjavia/netflix-movie-rating-dataset")


Please provide your Kaggle credentials to download this dataset. Learn more: http://bit.ly/kaggle-creds
Your Kaggle username: rahmaeladl
Your Kaggle Key: ··········
Dataset URL: https://www.kaggle.com/datasets/rishitjavia/netflix-movie-rating-dataset


100%|██████████| 74.7M/74.7M [00:01<00:00, 56.8MB/s]


In [ ]:
ratings=pd.read_csv("/content/netflix-movie-rating-dataset/Netflix_Dataset_Rating.csv")
ratings

User_ID  Rating  Movie_ID
0          712664       5         3
1         1331154       4         3
2         2632461       3         3
3           44937       5         3
4          656399       4         3
...           ...     ...       ...
17337453   520675       3      4496
17337454  1055714       5      4496
17337455  2643029       4      4496
17337456  1559566       3      4496
17337457   293198       3      4496

[17337458 rows x 3 columns]

In [ ]:
# counting unique values of movie_id
n = len(pd.unique(ratings['Movie_ID']))

print("No.of.unique values :", n)

No.of.unique values : 1350


In [ ]:
movies=pd.read_csv("/content/netflix-movie-rating-dataset/Netflix_Dataset_Movie.csv")
movies

Movie_ID  Year                                               Name
0             1  2003                                    Dinosaur Planet
1             2  2004                         Isle of Man TT 2004 Review
2             3  1997                                          Character
3             4  1994                       Paula Abdul's Get Up & Dance
4             5  2004                           The Rise and Fall of ECW
...         ...   ...                                                ...
17765     17766  2002  Where the Wild Things Are and Other Maurice Se...
17766     17767  2004                  Fidel Castro: American Experience
17767     17768  2000                                              Epoch
17768     17769  2003                                        The Company
17769     17770  2003                                       Alien Hunter

[17770 rows x 3 columns]

#Merge the 2 dataframes

In [ ]:
movies_ratings=pd.merge(ratings,movies)
movies_ratings


User_ID  Rating  Movie_ID  Year                   Name
0          712664       5         3  1997              Character
1         1331154       4         3  1997              Character
2         2632461       3         3  1997              Character
3           44937       5         3  1997              Character
4          656399       4         3  1997              Character
...           ...     ...       ...   ...                    ...
17337453   520675       3      4496  1993  Farewell My Concubine
17337454  1055714       5      4496  1993  Farewell My Concubine
17337455  2643029       4      4496  1993  Farewell My Concubine
17337456  1559566       3      4496  1993  Farewell My Concubine
17337457   293198       3      4496  1993  Farewell My Concubine

[17337458 rows x 5 columns]

#Some Information about merged dataframe

In [ ]:
#computing number of rows
rows = len(movies_ratings.axes[0])
print("Number of Rows: ", rows)

# computing number of columns
cols = len(movies_ratings.axes[1])
print("Number of Columns: ", cols)

Number of Rows:  17337458
Number of Columns:  5


#How many unique movies we have:  we have 1342 unique movies




In [ ]:
# counting unique values of movie_id/Name in merged df (movies_ratings)
unique_movies_names = len(pd.unique(movies_ratings['Name']))
print("No.of.unique movies :", unique_movies_names)

No.of.unique movies : 1342


#How many unique user id we have in merged dataframe (movies_ratings):  

In [ ]:
unique_users = len(pd.unique(movies_ratings['User_ID']))
print("No.of.unique users :", unique_users)

No.of.unique users : 143458


#Therefore if we transformed it to pivot table (its index is the users_id & (columns are the unique movies)  ..... the size of the pivot table is (143458 rows × 1342 columns)

#Transform it to pivot table:

In [ ]:
movies_ratings=movies_ratings.pivot_table(index=["User_ID"],columns=["Name"],values="Rating")
movies_ratings

Name     10  10 Things I Hate About You  \
User_ID                                   
6       NaN                         NaN   
7       NaN                         NaN   
79      NaN                         NaN   
97      NaN                         NaN   
134     NaN                         5.0   
...      ..                         ...   
2649370 NaN                         NaN   
2649378 NaN                         NaN   
2649388 NaN                         NaN   
2649426 NaN                         NaN   
2649429 NaN                         NaN   

Name     101 Dalmatians II: Patch's London Adventure  11:14  13 Ghosts  \
User_ID                                                                  
6                                                NaN    NaN        NaN   
7                                                NaN    NaN        NaN   
79                                               3.0    NaN        NaN   
97                                               NaN    NaN        NaN   
134                                              NaN    NaN        NaN   
...                                              ...    ...        ...   
2649370                                          NaN    NaN        NaN   
2649378                                          NaN    NaN        NaN   
2649388                                          NaN    NaN        NaN   
2649426                                          NaN    NaN        NaN   
2649429                                          NaN    NaN        NaN   

Name     18 Again  1984  2 Fast 2 Furious  200 Cigarettes  \
User_ID                                                     
6             NaN   4.0               NaN             NaN   
7             NaN   NaN               3.0             NaN   
79            NaN   NaN               2.0             NaN   
97            NaN   NaN               NaN             NaN   
134           NaN   NaN               5.0             NaN   
...           ...   ...               ...             ...   
2649370       NaN   NaN               NaN             NaN   
2649378       NaN   NaN               NaN             NaN   
2649388       NaN   NaN               NaN             NaN   
2649426       NaN   NaN               NaN             NaN   
2649429       NaN   NaN               NaN             NaN   

Name     2010: The Year We Make Contact  ...  \
User_ID                                  ...   
6                                   NaN  ...   
7                                   NaN  ...   
79                                  NaN  ...   
97                                  NaN  ...   
134                                 NaN  ...   
...                                 ...  ...   
2649370                             NaN  ...   
2649378                             NaN  ...   
2649388                             NaN  ...   
2649426                             NaN  ...   
2649429                             NaN  ...   

Name     Xena: Warrior Princess: Season 3  \
User_ID                                     
6                                     NaN   
7                                     NaN   
79                                    NaN   
97                                    NaN   
134                                   NaN   
...                                   ...   
2649370                               NaN   
2649378                               NaN   
2649388                               NaN   
2649426                               NaN   
2649429                               NaN   

Name     Xena: Warrior Princess: Series Finale  Y Tu Mama Tambien  \
User_ID                                                             
6                                          NaN                NaN   
7                                          NaN                NaN   
79                                         NaN                NaN   
97                                         3.0                NaN   
134                                        NaN            

In [ ]:
type(movies_ratings)

pandas.core.frame.DataFrame

In [ ]:
movies_ratings.columns

Index(['10', '10 Things I Hate About You',
       '101 Dalmatians II: Patch's London Adventure', '11:14', '13 Ghosts',
       '18 Again', '1984', '2 Fast 2 Furious', '200 Cigarettes',
       '2010: The Year We Make Contact',
       ...
       'Xena: Warrior Princess: Season 3',
       'Xena: Warrior Princess: Series Finale', 'Y Tu Mama Tambien',
       'Yellow Submarine', 'Yi Yi', 'Yojimbo', 'Young Black Stallion',
       'Youngblood', 'Yu-Gi-Oh!: The Movie', 'Zorro'],
      dtype='object', name='Name', length=1342)

#Search for all movies contains word (Wild) for example in (movies) dataframe

#I have 2 ways to make this :
#1- using for loop to iterate over movies['Name'] series

In [ ]:
wild_movies=[val for val in movies['Name'] if 'Wild' in val ]
print(wild_movies) #It is a list
print("number of movies contains wild word:  ",len(wild_movies))

['Lost in the Wild', 'Winston Churchill: The Wilderness Years', 'Wild Things', 'Reap the Wild Wind', "National Lampoon's Van Wilder", "Blue's Clues: Blue's Room: Beyond Your Wildest Dreams!", 'Wildflowers', 'The Wild Angels', 'Wild Style', "Jane Goodall's Wild Chimpanzees: IMAX", 'Mystery Science Theater 3000: The Wild World of Batwoman', 'How to Stuff a Wild Bikini', 'White Wolves: A Cry in The Wild 2', 'A Cry in the Wild', 'Wildboyz: Season 2', 'Born to be Wild', 'Wild in the Country', 'Wild Wild West', 'Simpsons Gone Wild', 'Nikki: Wild Dog of the North', 'Born to be Wild: The Leading Men of American Ballet Theatre', "Wacky Babies & Tender Times: Marty Stouffer's Wild America", 'Wild at Heart', 'The Wild Thornberrys Movie', 'Wilderness', 'Wildflower', 'Wildcats', 'Alaska: Spirit of the Wild: IMAX', 'Call of the Wild', 'White Wolves 2: Legend of The Wild', 'Due South: Call of the Wild', 'The Oscar Wilde Collection: Stage Right', 'The River Wild', 'In Desert and Wilderness', 'Touching

#2-Using -- Series.str.contains("specific word") - which returns true/false depending on occurance of the "specific word" in the values of the series.

https://www.geeksforgeeks.org/python-pandas-series-str-contains/

In [ ]:
wild_movies=movies['Name'][movies['Name'].str.contains("Wild")]
print(wild_movies) #It is a dataframe (it shows the index too)
print("number of movies contains wild word:  ",len(wild_movies))
print(type(movies_ratings))

139                                       Lost in the Wild
219                Winston Churchill: The Wilderness Years
329                                            Wild Things
841                                     Reap the Wild Wind
1508                         National Lampoon's Van Wilder
                               ...                        
17374    Ocean Wilds: Realm of the Killer Whales/Sperm ...
17390                                   Wildboyz: Season 1
17482    Crocodile Hunter: Wildest Home Videos / Big Cr...
17614                                  Scarecrow Gone Wild
17765    Where the Wild Things Are and Other Maurice Se...
Name: Name, Length: 76, dtype: object
number of movies contains wild word:   76
<class 'pandas.core.frame.DataFrame'>


#Display the first three values

In [ ]:
wild_movies.head(3)

139                           Lost in the Wild
219    Winston Churchill: The Wilderness Years
329                                Wild Things
Name: Name, dtype: object

#I want to see rating of users for one of the movies .. for example (Wild Things) movie..So, I will check movies_ratings dataframe

In [ ]:
wild_Things_ratings=movies_ratings['Wild Things']
wild_Things_ratings.head(10)

User_ID
6      NaN
7      NaN
79     NaN
97     NaN
134    5.0
169    NaN
183    NaN
188    NaN
195    NaN
199    4.0
Name: Wild Things, dtype: float64

#LAST STEP:

#I need to check correlation between ratings of this movie (Wild Things) and ratings of other movies

In [ ]:
correlation=movies_ratings.corrwith(wild_Things_ratings)
correlation.size

1342

In [ ]:
Similar_movies=correlation.sort_values(ascending=False)

In [ ]:
Similar_movies

Name
Wild Things                       1.000000
Poison Ivy 2                      0.441186
Players Club                      0.417944
Disturbing Behavior               0.384643
Jade                              0.382871
                                    ...   
The Shop Around the Corner       -0.031849
The Saddest Music in the World   -0.037492
Cries and Whispers               -0.042256
In the Mood for Love             -0.053918
The Weather Underground          -0.061764
Length: 1342, dtype: float64

#Conslusion:
---



# ** We conclude that the most 3 similar movies to (Wild Things ) movie -we can recommend-  are **
                    

#*  Poison Ivy 2  
#*  Players Club
#*  Disturbing Behavior**